src/State/Provider/ContentNegotiationProvider.php

```diff
@@ -108,14 +108,14 @@ private function getInputFormat(HttpOperation $operation, Request $request): ?st
             return $format;
         }
 
-        $supportedMimeTypes = [];
-        foreach ($formats as $mimeTypes) {
-            foreach ($mimeTypes as $mimeType) {
-                $supportedMimeTypes[] = $mimeType;
+        if (false !== $operation->canDeserialize() && !$request->isMethodSafe() && 'DELETE' !== $request->getMethod()) {
+            $supportedMimeTypes = [];
+            foreach ($formats as $mimeTypes) {
+                foreach ($mimeTypes as $mimeType) {
+                    $supportedMimeTypes[] = $mimeType;
+                }
             }
-        }
 
-        if (!$request->isMethodSafe() && 'DELETE' !== $request->getMethod()) {
             throw new UnsupportedMediaTypeHttpException(\sprintf('The content-type "%s" is not supported. Supported MIME types are "%s".', $contentType, implode('", "', $supportedMimeTypes)));
         }
 
```

Comment:
Would it be possible to not set input formats in
https://github.com/api-platform/core/blob/3.4/src/Metadata/Resource/Factory/FormatsResourceMetadataCollectionFactory.php#L79

when deserialize is `false`? Or do you think that the problem should be handled here? In that case can you check https://gist.github.com/soyuka/beac00bc527d2ea6e703ed27dd67532c/revisions (first is with the suggested patch, the second if we can move the handling of the formats to FormatsResourceMetadataCollectionFactory instead as there's no point having input formats if we don't deserialize imo)

Reply:
We need to check if the operation has input in the `ContentNegotiationProvider` even if we do not set the input format in `FormatsResourceMetadataCollectionFactory`.

So, to avoid overcomplicating the code and possibly not introducing some bugs or BC, I prefer not to change the `FormatsResourceMetadataCollectionFactory` logic.

And I apply your suggestion.